File: Task4/src/Predicate.scala
import scala.collection.mutable.ArrayBuffer

/**
 * Created by Шолохов on 30.01.2015.
 */
case class Predicate(name : String, terms : ArrayBuffer[Expression]) extends Expression {
  override def evaluate(values: Map[String, Boolean]): Boolean = true
  override def toString(): String = {
    var s : String = name
    if (terms != null && terms.size != 0) {
      s += "("
    }
    if (terms != null && terms.size != 0) {
      terms.foreach(t => s += t + ",")
      //s = s.dropRight(1)
    }
    if (terms != null && terms.size != 0) {
      s += ")"
    }
    s
  }

  override def equals(o : Any): Boolean = {
    o match {
      case t : Predicate =>
        var res = name.equals(t.name) && terms.size == t.terms.size
        if (res) {
          terms.zip(t.terms).foreach(x => res = res && x._1.equals(x._2))
        }
        res
      case _ => false
    }
  }
}

File: Task4/src/Predicate.java
import java.util.HashMap;
import java.util.ArrayList;

public class Predicate extends Statement {
    public final String name;
    public final ArrayList<Expression> terms;

    public Predicate(String name, ArrayList<Expression> terms) {
        this.name = name;
        this.terms = terms;
    }

    protected boolean compareImpl(String[] patterns, Statement other) {
        if (other instanceof Predicate) {
            Predicate pOther = (Predicate) other;
            if (name.equals(pOther.name) && terms.size() == pOther.terms.size()) {
                for (int i = 0; i < terms.size(); i++) {
                    if (!terms.get(i).equals(pOther.terms.get(i))) {
                        return false;
                    }
                }
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    public Statement substituteTerm(Expression haystack, Expression needle) {
        ArrayList<Expression> newTerms = new ArrayList<>();

        for (Expression term : terms) {
            newTerms.add(term.substituteTerm(haystack, needle));
        }

        return new Predicate(name, newTerms);
    }

    public Statement substitutePatterns(Statement[] to) {
        return this;
    }

    public boolean estimate(HashMap<String, Boolean> values) {
        return false;
    }

    public String toString() {
        String str = name;
        if (terms.size() != 0) {
            str += "(";
            for (int i = 0; i < terms.size(); i++) {
                str += terms.get(i);
                if (i != terms.size() - 1) {
                    str += ", ";
                }
            }
            str += ")";
        }
        return str;
    }
}
